python/src/iceberg/avro/file.py

```diff
@@ -0,0 +1,160 @@
+from iceberg.avro.reader import AvroStruct, ConstructReader
+from iceberg.io.base import InputFile, InputStream
+from iceberg.schema import Schema, visit
+from iceberg.types import (
+    BinaryType,
+    FixedType,
+    MapType,
+    NestedField,
+    StringType,
+    StructType,
+)
+from iceberg.utils.schema_conversion import AvroSchemaConversion
+
+VERSION = 1
+MAGIC = bytes(b"Obj" + bytearray([VERSION]))
+MAGIC_SIZE = len(MAGIC)
+SYNC_SIZE = 16
+META_SCHEMA = StructType(
+    NestedField(name="magic", field_id=100, field_type=FixedType(length=MAGIC_SIZE), is_optional=False),
+    NestedField(
+        field_id=200,
+        name="meta",
+        field_type=MapType(key_id=201, key_type=StringType(), value_id=202, value_type=BinaryType(), value_is_optional=False),
+        is_optional=False,
+    ),
+    NestedField(field_id=300, name="sync", field_type=FixedType(length=SYNC_SIZE), is_optional=False),
+)
+
+
+@dataclass(frozen=True)
+class AvroFileHeader:
+    magic: bytes
+    meta: dict[str, str]
+    sync: bytes
+
+    def get_compression_codec(self) -> type[Codec]:
+        """Get the file's compression codec algorithm from the file's metadata."""
+        codec_key = "avro.codec"
+        if codec_key in self.meta:
+            codec_name = self.meta[codec_key]
+
+            if codec_name not in KNOWN_CODECS:
+                raise ValueError(f"Unsupported codec: {codec_name}. (Is it installed?)")
+
+            return KNOWN_CODECS[codec_name]
+        else:
+            return NullCodec
+
+    def get_schema(self) -> Schema:
+        schema_key = "avro.schema"
+        if schema_key in self.meta:
+            avro_schema_string = self.meta[schema_key]
+            avro_schema = json.loads(avro_schema_string)
+            return AvroSchemaConversion().avro_to_iceberg(avro_schema)
+        else:
+            raise ValueError("No schema found in Avro file headers")
+
+
+class AvroFile:
+    input_file: InputFile
+    input_stream: InputStream
+    header: AvroFileHeader
+    schema: Schema
+    file_length: int
+
+    decoder: BinaryDecoder
+    block_decoder: BinaryDecoder
+
+    block_records: int = -1
+    block_position: int = 0
+    block: int = 0
+
+    def __init__(self, input_file: InputFile) -> None:
+        self.input_file = input_file
+
+    def __enter__(self):
+        """
+        Opens the file and reads the header and generates
+        a reader tree to start reading the payload
+
+        Returns:
+            A generator returning the AvroStructs
+        """
+        self.input_stream = self.input_file.open()
+        self.decoder = BinaryDecoder(self.input_stream)
+        self.header = self._read_header()
+        self.schema = self.header.get_schema()
+        self.file_length = len(self.input_file)
+        self.reader = visit(self.schema, ConstructReader())
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.input_stream.close()
+
+    def __iter__(self) -> AvroFile:
```

Comment:
Is this equivalent to `iterator` in Java? If so, then structuring this class to return `self` rather than a reader means that you can't share the instance. That's not a problem, but the best practice in Java is to use a separate `Iterator` class that has its own open stream and state. I'm wondering if that's the best practice in Python as well?

Reply:
I've combined Java's closable and the iterator on this one. I could refactor this one a bit, for example by for clarity by splitting out the block